### PR TITLE
PORT-6978 | DORA Metrics Mini Guides

### PR DESCRIPTION
**Pull request by mk-armah**
# Description

- Added a new guide (`/connect-dora-metrics-to-service`)

- Updated DORA Metrics guide with team metrics
- Adds new properties (`Response Rate` and `Average Response Time` and `Time Frame`) to github blueprint to host team metrics in port.
- Computes response rate and response time for github teams
- Upserts metrics alongside port github blueprint properties

image

image

## Added docs pages

A new guide for connecting dora metrics to a service

- Connect DORA Metrics to Service (`/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service`)

## Updated docs pages

This update introduces the integration of team metrics to measure the response rate and response time of review requests.

#### Key Features:

1. **Response Rate**
   - **Definition**: Percentage of PRs responded to by team members.
   - **Scope**: PRs created within the specified time frame.
   - **Criteria**: PRs requesting a review from the team.
   - **Formula**: 
     - Response Rate = (Number of PRs reviewed by team members / Total PRs requesting team review) * 100

2. **Response Time**
   - **Definition**: Time taken by team members to respond to PRs.
   - **Scope**: PRs created within the specified time frame and reviewed by team members.
   - **Criteria**: First review by a team member for each PR.
   - **Formula**: 
     - Average Response Time = Total response time for all counted reviews / Number of counted reviews

3. Support for Github self hosted instances
---

- DORA Metrics (`/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics`)

### CODE REVIEW

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md [hunk 1]
```diff
@@ -6,20 +6,233 @@ import PortTooltip from "/src/components/tooltip/tooltip.jsx";
...
 
 
 ## Prerequisites
 1. A GitHub repository in which you can trigger a workflow that we will use in this guide.
-2. A blueprint in Port to host the Dora Metrics.
+2. Create the following GitHub action secrets:
 
+    - `PORT_CLIENT_ID` - [Port Client ID](/build-your-software-catalog/custom-integration/api/#get-api-token)
+    - `PORT_CLIENT_SECRET` - [Port Client Secret](/build-your-software-catalog/custom-integration/api/#get-api-token)
+    - `GH_PAT_CLASSIC` - [GitHub Personal Access Token (Classic)](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#personal-access-tokens-classic). Ensure that the `read:org` and `repo`, scopes are set for the token to grant this action access to the repositories and teams the metrics are to be estimated for . [learn more].
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md:18]: The learn more here does not link anywhere

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md [hunk 2]
```diff
@@ -6,20 +6,233 @@ import PortTooltip from "/src/components/tooltip/tooltip.jsx";
...
+          python-version: '3.x'
+
+      - name: Install Python dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r dora/requirements.txt
+
+      - name: Compute PR Metrics
+        run: |
+          python dora/calculate_pr_metrics.py  --owner "${{ needs.setup.outputs.owner }}" --repo "${{ matrix.repository }}" --token "${{ secrets.GH_PAT_CLASSIC }}" --time-frame "${{ needs.setup.outputs.doraTimeFrame }}" --platform github-actions --github-host "${{ needs.setup.outputs.githubHost }}"
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md:123]: I don't see `calculate_pr_metrics.py` anywhere else in the doc, please verify the file names

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md [hunk 3]
```diff
@@ -6,20 +6,233 @@ import PortTooltip from "/src/components/tooltip/tooltip.jsx";
...
+httpx==0.27.0
+```
+
+</details>
+
+
+<details>
+  <summary>Workflow Configuration</summary>
+:::tip
+You can choose to run any of the metric jobs (`compute-team-metrics` or `compute-repo-metrics`) optionally. This is controlled by setting or unsetting the corresponding Port blueprint parameters in your configuration file (`dora/dora-config.json`)
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md:185]: This is unclear, after reading this tip, I still do not know how to correctly control what is computed, I think the table below needs to be expanded to explain it better. Or you can provide an additional example `dora-config.json` which shows some other formats

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md [hunk 4]
```diff
@@ -6,20 +6,233 @@ import PortTooltip from "/src/components/tooltip/tooltip.jsx";
...
+PyGithub==2.3.0
+httpx==0.27.0
+```
+
+</details>
+
+
+<details>
+  <summary>Workflow Configuration</summary>
+:::tip
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md:184]:
```suggestion
:::tip Controlling configurations
```

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md [hunk 5]
```diff
@@ -813,22 +917,442 @@ if __name__ == "__main__":
...
 
-Congrats 🎉 You've successfully created a scheduled GitHub workflow that periodically calculates and ingests `DORA Metrics` for GitHub repository(s).
+## Team Metrics
+
+1. Add the following new properties, `averageResponseTime`, `averageResponseTime`,and `timeFrame` to your GitHub Team blueprint in port
 
+<details>
+<summary> GitHub Team blueprint (click to expand) </summary>
+
+```json showLineNumbers title="github team blueprint" {29,30,31,32,33,34,35,36,37,38,39,40,41}
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md:949]: Please use `// highlight-start` and `// highlight-end` as we do in other code blocks in the docs, and don't use the {1,2,3}

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import PortTooltip from "/src/components/tooltip/tooltip.jsx"
+
+# Connect DORA Metrics to Service
+
+This guide aims to demonstrate how to connect DORA Metrics to a service in Port.
+
+:::tip Prerequisites
+
+- This guide assumes you have a Port account and that you have finished the [onboarding process](/quickstart).
+- Install port [GitHub integration](/build-your-software-catalog/sync-data-to-catalog/git/github/installation)
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service.md:10]:
```suggestion
- Install Port's [GitHub integration](/build-your-software-catalog/sync-data-to-catalog/git/github/installation) 

```

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    entity:
+      mappings:
+        identifier: .name
+        title: .name
+        blueprint: '"service"'
+        properties:
+          readme: file://README.md
+          url: .html_url
+          language: .language
+        relations:
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service.md:56]: Add a highlight here to show that this is what changed

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+          doraMetrics: .id
+```
+</details>
+
+What we just did was map the `DORA Metrics` to the relation between it and our `Services`.  
+Now, if our `DORA Metrics` identifier is equal to the repository id, the `metrics` will automatically be linked to it &nbsp;🎉
```
Comment (MPTG94) [inline docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/connect-dorametrics-to-service.md:62]:
```suggestion
Now, if our `DORA Metrics` identifier is equal to the repository id, the `metrics` will automatically be linked to it 🎉

```

**Review by MPTG94 — CHANGES_REQUESTED** [review]
Take a look at my comments

File: docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md (hunk 2 above)
Comment (mk-armah) [reply docs/build-your-software-catalog/custom-integration/api/ci-cd/github-workflow/guides/dora-metrics/dora-metrics.md:123]: Line 397 -590

**Pull request closed without merging**